### PR TITLE
BUG: fetch vram data from cluster

### LINKED ISSUE #327: ENH: `StorageAPI.get` supports convert GPU buffers into CPU buffers
### Is your feature request related to a problem? Please describe
The supervisor doesn't have any workload on it. Thus, it should have nothing to do with GPUs.

But currently, the supervisor needs a cuda device to receive buffers from workers when fetching results.

### Describe the solution you'd like
Add an option `cpu` to `StorageAPI.get` that make the returned buffers always CPU buffers.

### LINKED ISSUE #348: BUG:  fetching vram data from cluster
### Describe the bug
In a cluster deployment with GPU enabled, fetching data from supervisor via web storage API fails.

### To Reproduce

1. Your Python version: `3.9.15`
2. The version of Xorbits you use: `master`
3. Versions of crucial packages, such as numpy, scipy and pandas
4. Full stack of the error:
```python
Traceback (most recent call last):
  File "/home/wangzhong/repos/xorbits/benchmarks/tpch/run_queries.py", line 1141, in <module>
    main()
  File "/home/wangzhong/repos/xorbits/benchmarks/tpch/run_queries.py", line 1128, in main
    run_queries(
  File "/home/wangzhong/repos/xorbits/benchmarks/tpch/run_queries.py", line 1058, in run_queries
    globals()[f"q{query:02}"](*queries_to_args[query])
  File "/home/wangzhong/repos/xorbits/benchmarks/tpch/run_queries.py", line 122, in wrapped
    q(*args, **kwargs)
  File "/home/wangzhong/repos/xorbits/benchmarks/tpch/run_queries.py", line 191, in q01
    print(total)
  File "/home/wangzhong/repos/xorbits/python/xorbits/utils.py", line 37, in inn
    return f(self, *args, **kwargs)
  File "/home/wangzhong/repos/xorbits/python/xorbits/core/data.py", line 243, in __str__
    return self.data.__str__()
  File "/home/wangzhong/repos/xorbits/python/xorbits/core/data.py", line 98, in __str__
    return self._mars_entity.__str__()
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/core/entity/core.py", line 102, in __str__
    return self._data.__str__()
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/dataframe/core.py", line 2180, in __str__
    return self._to_str(representation=False)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/dataframe/core.py", line 2152, in _to_str
    corner_data = fetch_corner_data(self, session=self._executed_sessions[-1])
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/dataframe/utils.py", line 1158, in fetch_corner_data
    return df_or_series._fetch(session=session)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/core/entity/executable.py", line 169, in _fetch
    return fetch(self, session=session, **kw)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1928, in fetch
    return session.fetch(tileable, *tileables, **kwargs)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1707, in fetch
    return asyncio.run_coroutine_threadsafe(coro, self._loop).result()
  File "/home/wangzhong/miniconda3/envs/gpu/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/home/wangzhong/miniconda3/envs/gpu/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1896, in _fetch
    data = await session.fetch(tileable, *tileables, **kwargs)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1125, in fetch
    fetched_data = await fetcher.get()
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/services/task/execution/mars/fetcher.py", line 52, in get
    fetched_data = await storage_api.get.batch(
  File "/home/wangzhong/miniconda3/envs/gpu/lib/python3.9/site-packages/xoscar/batch.py", line 147, in _async_batch
    return [await self._async_call(*args_list[0], **kwargs_list[0])]
  File "/home/wangzhong/miniconda3/envs/gpu/lib/python3.9/site-packages/xoscar/batch.py", line 96, in _async_call
    return await self.func(*args, **kwargs)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/services/storage/api/web.py", line 124, in get
    res = await self._request_url(
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/services/web/core.py", line 274, in _request_url
    raise exc.with_traceback(tb)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/services/web/core.py", line 69, in wrapped
    res = await func(self, *args, **kwargs)
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/services/storage/api/web.py", line 79, in get_data_by_post
    self.write(serialize_serializable(result))
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/utils.py", line 431, in serialize_serializable
    buf_sizes = [getattr(buf, "nbytes", len(buf)) for buf in buffers]
  File "/home/wangzhong/repos/xorbits/python/xorbits/_mars/utils.py", line 431, in <listcomp>
    buf_sizes = [getattr(buf, "nbytes", len(buf)) for buf in buffers]
TypeError: object of type 'Buffer' has no len()
```
6. Minimized code to reproduce the error:

Deploy an Xorbits cluster:
```bash
python -m xorbits.supervisor -H 0.0.0.0 -p 7777 -w 7778
```

```bash
python -m xorbits.worker -H 127.0.0.1 -p 7779 -s 127.0.0.1:7777 --cuda-devices 0
```

```bash
python /path/to/xorbits/benchmarks --data_set /path/to/tpch-sf1 --endpoint http://127.0.0.1:7778 --queries 1
```

### PR DESCRIPTION
**Pull request by UranusSeven**
## What do these changes do?
Add an option for `StorageAPI.get` to convert `cudf` data structures to `pandas` data structures.

## Related issue number
Resolves #348 
Resolves #327 

## Check code requirements

- [x] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/_mars/services/storage/handler.py [hunk 1]
```diff
@@ -124,13 +124,18 @@ async def get(
...
         conditions: List = None,
+        cpu: Optional[bool] = None,
         error: str = "raise",
     ):
         try:
             data_info = await self._data_manager_ref.get_data_info(
                 session_id, data_key, self._band_name
             )
             data = yield self._get_data(data_info, conditions)
+            from xorbits._mars.dataframe.utils import is_cudf
```
Comment (qianduoduo0904) [inline python/xorbits/_mars/services/storage/handler.py:135]: Could use relative path import here.

File: python/xorbits/_mars/services/storage/handler.py [hunk 2]
```diff
@@ -124,13 +124,18 @@ async def get(
...
+
+            if is_cudf(data) and cpu:
+                data = data.to_pandas()
```
Comment (qianduoduo0904) [inline python/xorbits/_mars/services/storage/handler.py:138]: Looks like no ut covered this line.

Comment (UranusSeven) [reply python/xorbits/_mars/services/storage/handler.py:138]: Yeah, I'm working on new test cases for this scenario.

File: python/xorbits/_mars/services/storage/handler.py (hunk 1 above)
Comment (UranusSeven) [reply python/xorbits/_mars/services/storage/handler.py:135]: Good catch.

File: python/xorbits/_mars/services/storage/tests/test_service.py [hunk 3]
```diff
@@ -186,25 +186,33 @@ async def test_storage_service_with_cuda(actor_pools_with_gpu):
...
+            "col3": np.random.rand(10),
+        },
     )
+    data2 = cudf.DataFrame(data2_pd)
     await storage_api.put("mock_cudf_key", data2, level=StorageLevel.GPU)
     get_data2 = await storage_api.get("mock_cudf_key")
     assert isinstance(get_data2, cudf.DataFrame)
     cudf.testing.assert_frame_equal(data2, get_data2)
+    get_data2_pd = await storage_api.get("mock_cudf_key", cpu=True)
+    print(type(get_data2_pd))
```
Comment (aresnow1) [inline python/xorbits/_mars/services/storage/tests/test_service.py:213]: Remove the code for debugging.

File: python/xorbits/_mars/services/storage/tests/test_service.py [hunk 4]
```diff
@@ -186,25 +186,33 @@ async def test_storage_service_with_cuda(actor_pools_with_gpu):
...
-    data2 = cudf.DataFrame(
-        pd.DataFrame(
-            {
-                "col1": np.arange(10),
-                "col2": [f"str{i}" for i in range(10)],
-                "col3": np.random.rand(10),
-            },
-        )
+    get_data1_np = await storage_api.get("mock_cupy_key", cpu=True)
+    print(type(get_data1_np))
```
Comment (aresnow1) [inline python/xorbits/_mars/services/storage/tests/test_service.py:196]: Remove the code for debugging.

File: python/xorbits/_mars/services/storage/api/oscar.py [hunk 5]
```diff
@@ -93,10 +93,14 @@ async def is_seekable(self, storage_level: StorageLevel = None) -> bool:
 
     @mo.extensible
     async def get(
-        self, data_key: str, conditions: List = None, error: str = "raise"
+        self,
+        data_key: str,
+        conditions: List = None,
+        cpu: Optional[bool] = None,
```
Comment (aresnow1) [inline python/xorbits/_mars/services/storage/api/oscar.py:99]: `cpu` is confusing for me, but I don't think of a better name.

**Review by qianduoduo0904 — DISMISSED** [review]
LGTM

File: python/xorbits/_mars/services/storage/api/oscar.py (hunk 5 above)
Comment (qianduoduo0904) [reply python/xorbits/_mars/services/storage/api/oscar.py:99]: Maybe `to_cpu`? but `cpu` is OK to me.

Comment (UranusSeven) [reply python/xorbits/_mars/services/storage/api/oscar.py:99]: The meaning of this parameter is moving data in vram to ram, so how about `to_main_mem`?

Comment (aresnow1) [reply python/xorbits/_mars/services/storage/api/oscar.py:99]: `to_cpu` is OK.

**Review by aresnow1 — APPROVED** [review]
LGTM

**Pull request merged**